### PR TITLE
feat: Add URL option to testing script

### PR DESCRIPTION
**Pull request by coolalexzb**
`run.py ` is a Python script that performs a hypothesis test using `pyhf` package automatically.
This PR aims to complete the automatic downloading process, which is the first part of the whole process.
Current version is built on `example.py`
All the data downloaded online will be stored in `data` directory.
More tests are needed.

### CODE REVIEW

**Comment by matthewfeickert** [comment]
> All the data downloaded online will be stored in `data` directory.

This is fine, but why are the JSON files being put in version control here? They are already versioned and publicly available, so it seems strange to have duplicate version of a million line file. Probably better to just provide methods to _get_ the JSON.

File: src/run.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    response = requests.get(url, stream=True)
+    assert response.status_code == 200
+    with open(targz_filename, "wb") as file:
+        file.write(response.content)
+
+    # Open as a tarfile
+    tar = tarfile.open(targz_filename, "r:gz")
+    filenames = tar.getnames()
+    tar.extractall(path="../data/")
```
Comment (matthewfeickert) [inline src/run.py]: Probably better to use [`pathlib`](https://docs.python.org/3/library/pathlib.html) for all paths to make it OS independent by default.

**Comment by coolalexzb** [comment]
Since the users have the option of using local source data. If we can store data locally, the users can use them next time. Besides, it can maintain the same type of return value of function `open_local_file` and function `downlaod`, which makes the code clean. 

If you don't hope to store online data locally, I can make a change.

**Comment by coolalexzb** [comment]
Another point: is it possible that the downloaded data will be formats like tar, zip? The current code only support tar.gz file

**Comment by matthewfeickert** [comment]
> Since the users have the option of using local source data. If we can store data locally, the users can use them next time.

I'm not against the idea of using local data &mdash; not everything needs to be downloaded or streamed everytime something is run. But I don't see the reason to duplicate files like this and to keep them under version control with the source for the benchmarking. A user should be able to download them locally and keep them with this download code, but why put it under version control in this repo?

> Another point: is it possible that the downloaded data will be formats like tar, zip? The current code only support tar.gz file

No, HEPData only provides `tar.gz`. That's fine for our use cases.

**Comment by coolalexzb** [comment]
I just hope that users can use the data under version control to test loading local data. If it is not necessary, I think I can delete it.

**Comment by matthewfeickert** [comment]
> users can use the data under version control to test loading local data

Okay sure, but if this is for test purposes then this should be arbitrary test JSON that is put in with the unit tests. We shouldn't be duplicating a million line JSON file that already publicly exists just to run a use test.

**Comment by coolalexzb** [comment]
Got it! I also mentioned two questions on Slack. I will resolve all the problems after a one-hour meeting. Thank you!

File: src/run.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    "-u",
+    "--url",
+    "url",
+    help="Online data link.",
+    default="https://www.hepdata.net/record/resource/1267798?view=true",
+    required=False,
+)
+@click.option(
+    "-m",
+    "--model_point",
```
Comment (matthewfeickert) [inline src/run.py]:
```suggestion
    "--model-point",
```

For a CLI it is probably nicer to user dashes over underscores, but should still use underscores when actually in the code (like you're doing now).

File: src/run.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    "model_point",
+    help="Model point.",
+    default=(750, 100),
```
Comment (matthewfeickert) [inline src/run.py]:
```suggestion
```

Probably don't want a default here

File: src/run.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    else:
+        print("Invalid command!")
+        print("Command help ....")
+        print(
+            "python run.py -b numpy -u https://www.hepdata.net/record/resource/1267798?view=true -m (750, 100)"
+        )
+        print(
+            "python run.py -b numpy -p ../data/1Lbb-likelihoods-hepdata -m (750, 100)"
+        )
+        raise
```
Comment (matthewfeickert) [inline src/run.py]: Probably should avoid using bare `raise` statements, even if it is for a custom error.

File: src/run.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        directory_name, model_point,
+    )
+
+    print("\nStarting fit\n")
+    fit_start_time = datetime.now()
+    CLs_obs, CLs_exp = calculate_CLs(background, signal_patch)
+    fit_end_time = datetime.now()
+    fit_time = fit_end_time - fit_start_time
+
+    print(f"fit C1N2_Wh_hbb_750_100 in {fit_time} seconds\n")
```
Comment (matthewfeickert) [inline src/run.py]: Don't want the analysis (C1N2_Wh_hbb_750_100) hardcoded in here

File: src/run.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    help="Local path of workspace.",
+    default=None,
+    required=False,
+)
+@click.option(
+    "-u",
+    "--url",
+    "url",
+    help="Online data link.",
+    default="https://www.hepdata.net/record/resource/1267798?view=true",
```
Comment (matthewfeickert) [inline src/run.py]:
```suggestion
```
Probably don't want a default here.

**Review by matthewfeickert — COMMENTED** [review]
I think this is looking good. :+1: I just have a few comments and then a suggestion that this not be located under `src/` but under something like `scripts/`. The reason I suggest moving out from under `src/` is that we're eventually going to want to turn this into an actual library, and we'll want to use a modern `src/` directory structure for that. (c.f. the `pyhf` repo structure, or watch this [video from Anthony Sottile](https://youtu.be/sW1qUZ_nSXk) on why to use `src/`)

File: src/run.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    result = pyhf.infer.hypotest(
+        1.0, workspace.data(model), model, qtilde=True, return_expected_set=True
+    )
+    if isinstance(pyhf.tensorlib, pyhf.tensor.pytorch_backend):
+        return result[0].tolist()[0], result[-1].tolist()
+    else:
+        return result[0].tolist()[0], result[-1].ravel().tolist()
+
+def end(directory_name):
```
Comment (matthewfeickert) [inline src/run.py]: oh, also given the name I assumed that this was some Python builtin that I hadn't heard of before (when I saw it down below first). Probably better to give it a more descriptive name that says what it actually does.

File: src/run.py (hunk 3 above)
Comment (coolalexzb) [reply src/run.py]: Without a default model point here. The script cannot function well. 
```
result = pyhf.infer.hypotest(
        1.0, workspace.data(model), model, qtilde=True, return_expected_set=True
    )
```
model(model point) is a parameter required here
In the script, I expect to use `C1N2_Wh_hbb_750_100` as a default input

File: src/run.py (hunk 6 above)
Comment (coolalexzb) [reply src/run.py]: So, we don't want to use a default value for URL or path? The reason why I add default values here is that I just hope that I can run `run.py` in Pycharm without using a command line.

Comment (matthewfeickert) [reply src/run.py]: Default values are added if they make sense for the end user, which it doesn't as we want to be able to run on an arbitrary file. Can't you just run using the CLI in PyCharm's terminal emulator?

File: src/run.py (hunk 3 above)
Comment (matthewfeickert) [reply src/run.py]: Defaults only make sense if they are something that is independent of the files being looked at. Here a user should be providing information as they should know what they want to test if they want to run on a specific model point.

**Comment by matthewfeickert** [comment]
Oh, also if this PR can introduce some example usage of this script from the CLI to the `README` or some other Markdown document that would be great.

File: src/run.py (hunk 4 above)
Comment (coolalexzb) [reply src/run.py]: Is there any other more elegant way to remind the users of their wrong command?

Comment (matthewfeickert) [reply src/run.py]: I just meant that using `raise` with no clear indication of what exception you're raising should be avoided. https://docs.python.org/3/tutorial/errors.html#raising-exceptions

I see that you now just `return`, which works too.

File: scripts/run.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
+import requests
+import hashlib
```
Comment (matthewfeickert) [inline scripts/run.py]:
```suggestion
```

Removing the unused `hashlib` will make LGTM happy so that it will stop warning about this.

**Review by matthewfeickert — APPROVED** [review]
`scripts/__init__.py` doesn't need to exist as this isn't part of a package. https://docs.python.org/3/tutorial/modules.html#packages

But beyond that this LGTM. :+1: 

When you squash and merge it in though maybe tidy up commit log to just summarize what was done, as there are (for good reason) lots of commits here.

**Comment by coolalexzb** [comment]
Is this PR ready to be merged?

**Comment by matthewfeickert** [comment]
> Is this PR ready to be merged?

If you're happy with it then LGTM. :+1: Though if you had planned to add something else in the PR that I wasn't aware of then don't feel forced to merge early.

**Comment by coolalexzb** [comment]
I expected this PR can be used to complete the automatic download part of a testing script. Can I change the title of this PR back to automatic download and merge? I will complete and improve the rest parts of the testing script, such as computation, benchmark, in other PRs. Or I can maintain this PR and continue the rest part of the testing script?

**Comment by matthewfeickert** [comment]
> Can I change the title of this PR back to automatic download and merge? I will complete and improve the rest parts of the testing script, such as computation, benchmark, in other PRs. Or I can maintain this PR and continue the rest part of the testing script?

Sure. Either works, but I agree that trying to break things into smaller "atomic" PRs that just focus on one feature is probably the better way to go then trying to add in everything all at once in one PR.

**Pull request merged**